DirectProgramming/C++SYCL_FPGA/Tutorials/GettingStarted/fpga_compile/part2-dpcpp_functor_usm/src/vector_add.cpp

```diff
@@ -0,0 +1,96 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/sycl.hpp>
+#include <sycl/ext/intel/fpga_extensions.hpp>
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class VectorAddID;
+
+struct VectorAdd {
+    int *const vec_a_in;
+    int *const vec_b_in;
+    int *const vec_c_out;
+    int len;
+
+    void operator()() const {
+        for (int idx = 0; idx < len; idx++) {
+            int a_val  = vec_a_in[idx];
+            int b_val  = vec_b_in[idx];
+            int sum    = a_val + b_val;
+            vec_c_out[idx] = sum;
+        }
+    }
+};
+
+constexpr int kVectSize = 256;
+
+int main() {
+    bool passed = true;
+    try {
+        // Use compile-time macros to select either:
+        //  - the FPGA emulator device (CPU emulation of the FPGA)
+        //  - the FPGA device (a real FPGA)
+        //  - the simulator device 
+#if FPGA_SIMULATOR
+        auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+        auto selector = sycl::ext::intel::fpga_selector_v;
+#else  // #if FPGA_EMULATOR
+        auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+        // create the device queue
+        sycl::queue q(selector);
+
+        auto device = q.get_device();
+
+        std::cout << "Running on device: "
+                << device.get_info<sycl::info::device::name>().c_str()
+                << std::endl;
+
```

Comment:
```suggestion

if (!device.has(aspect::usm_host_allocations)) {
     std::terminate();
}


```

Check for USM support and exit gracefully